File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.java
import java.util.*;

class Solution {
    public String solution(int[] numbers, String hand) {
        List<List<Integer>> keypad = new ArrayList<>();
        keypad.add(List.of(1, 4, 7, -1));
        keypad.add(List.of(2, 5, 8, 0));
        keypad.add(List.of(3, 6, 9, -1));
        
        StringBuilder sb = new StringBuilder();
        int[] left = {0, 3}; int[] right = {2, 3};
        int idx, leftDist, rightDist;
        for (int n : numbers) {
            if (keypad.get(0).indexOf(n) != -1) {
                sb.append("L");
                left[0] = 0; left[1] = keypad.get(0).indexOf(n);
            }
            if (keypad.get(2).indexOf(n) != -1) {
                sb.append("R");
                right[0] = 2; right[1] = keypad.get(2).indexOf(n);
            }
            if (keypad.get(1).indexOf(n) != -1) {
                idx = keypad.get(1).indexOf(n);
                leftDist = Math.abs(1 - left[0]) + Math.abs(left[1] - idx);
                rightDist = Math.abs(1 - right[0]) + Math.abs(right[1] - idx);
                if (leftDist == rightDist) {
                    if (hand.equals("left")) {
                        sb.append("L");
                        left[0] = 1; left[1] = idx;
                    } else {
                        sb.append("R");
                        right[0] = 1; right[1] = idx;
                    }
                } else if (leftDist < rightDist) {
                    sb.append("L");
                    left[0] = 1; left[1] = idx;
                } else {
                    sb.append("R");
                    right[0] = 1; right[1] = idx;
                }
            }
        }
        return sb.toString();
    }
}

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.py
def solution(numbers, hand):
    keypad = [[1, 4, 7, -1], [2, 5, 8, 0], [3, 6, 9, -1]]
    
    answer = ''
    left, right = [0, 3], [2, 3] # [0] : 0, 1, 2 | [1] : 0, 1, 2, 3
    for n in numbers:
        if n in keypad[0]:
            answer += 'L'
            left = [0, keypad[0].index(n)]
        if n in keypad[2]:
            answer += 'R'
            right = [2, keypad[2].index(n)]
        if n in keypad[1]:
            idx = keypad[1].index(n)
            left_dist = abs(1 - left[0]) + abs(left[1] - idx)
            right_dist = abs(1 - right[0]) + abs(right[1] - idx)
            if left_dist == right_dist:
                if hand == 'left':
                    answer += 'L'
                    left = [1, idx]
                else:
                    answer += 'R'
                    right = [1, idx]
            elif left_dist < right_dist:
                answer += 'L'
                left = [1, idx]
            else:
                answer += 'R'
                right = [1, idx]
    return answer